src/backend/distributed/metadata/node_metadata.c

```diff
@@ -753,13 +754,662 @@ PgDistTableMetadataSyncCommandList(void)
+		{
+			forceDelegation = NO_FORCE_PUSHDOWN;
+		}
+		else
+		{
+			forceDelegation = DatumGetBool(forceDelegationDatum);
+		}
+
+		char *workerMetadataUpdateCommand =
+			MarkObjectsDistributedCreateCommand(list_make1(objectAddress),
+												list_make1_int(distArgumentIndex),
+												list_make1_int(colocationId),
+												list_make1_int(forceDelegation));
+		List *commandList = list_make1(workerMetadataUpdateCommand);
+		commandList = lcons(DISABLE_DDL_PROPAGATION, commandList);
+		commandList = lappend(commandList, ENABLE_DDL_PROPAGATION);
+		SendMetadataCommandListToWorkerListInCoordinatedTransaction(nodesWithMetadata,
+																	CurrentUserName(),
+																	commandList);
+	}
+
+	systable_endscan_ordered(scanDesc);
+	index_close(indexRelation, AccessShareLock);
+	table_close(relation, AccessShareLock);
+	list_free(nodesWithMetadata);
+
+	MemoryContextSwitchTo(oldContext);
+	MemoryContextDelete(metadataSyncContext);
+}
+
+
+/*
+ * SendColocationMetadataCommands sends colocation metadata within metadata
+ * connection.
+ */
+void
+SendColocationMetadataCommands(List *nodeList)
+{
+	List *nodesWithMetadata = NIL;
+	WorkerNode *workerNode = NULL;
+	foreach_ptr(workerNode, nodeList)
+	{
+		if (NodeIsPrimary(workerNode) && !NodeIsCoordinator(workerNode))
+		{
+			nodesWithMetadata = lappend(nodesWithMetadata, workerNode);
+		}
+	}
+
+	if (nodesWithMetadata == NIL)
+	{
+		return;
+	}
+
+	List *colocationGroupSyncCommandList = ColocationGroupCreateCommandList();
+
+	if (colocationGroupSyncCommandList == NIL)
+	{
+		return;
+	}
+
+	colocationGroupSyncCommandList = lcons(DISABLE_DDL_PROPAGATION,
+										   colocationGroupSyncCommandList);
+	colocationGroupSyncCommandList = lappend(colocationGroupSyncCommandList,
+											 ENABLE_DDL_PROPAGATION);
+	SendMetadataCommandListToWorkerListInCoordinatedTransaction(nodesWithMetadata,
+																CurrentUserName(),
+																colocationGroupSyncCommandList);
+	list_free(nodesWithMetadata);
+}
+
+
+/*
+ * SendNodeWideObjectsSyncCommands sends systemwide objcets to workers within
+ * metadata connection.
+ */
+void
+SendNodeWideObjectsSyncCommands(List *nodeList)
+{
+	List *nodesWithMetadata = NIL;
+	WorkerNode *workerNode = NULL;
+	foreach_ptr(workerNode, nodeList)
+	{
+		if (NodeIsPrimary(workerNode) && !NodeIsCoordinator(workerNode))
+		{
+			nodesWithMetadata = lappend(nodesWithMetadata, workerNode);
+		}
+	}
+
+	if (nodesWithMetadata == NIL)
+	{
+		return;
+	}
+
+	/*
+	 * Propagate node wide objects. It includes only roles for now.
+	 */
+	List *commandList = PropagateNodeWideObjectsCommandList();
+
+	if (commandList == NIL)
+	{
+		return;
+	}
+
+	commandList = lcons(DISABLE_DDL_PROPAGATION,
+						commandList);
+	commandList = lappend(commandList,
+						  ENABLE_DDL_PROPAGATION);
+	SendMetadataCommandListToWorkerListInCoordinatedTransaction(nodesWithMetadata,
+																CurrentUserName(),
+																commandList);
+	list_free(nodesWithMetadata);
```

Comment:
instead of list_free, lets create a memory context at the start of the function, and free at the end to prevent all potential leaks?

Reply:
I removed that in later commits.